ugs-platform/ugs-platform-plugin-cloud-storage/src/main/java/com/willwinder/ugs/nbp/S3FileSystemView.java

```diff
@@ -0,0 +1,340 @@
+/*
+    Copyright 2020 Will Winder
+
+    This file is part of Universal Gcode Sender (UGS).
+
+    UGS is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    UGS is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with UGS.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.willwinder.ugs.nbp;
+
+import io.minio.MinioClient;
+import io.minio.Result;
+import io.minio.messages.Bucket;
+import io.minio.messages.Item;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.swing.Icon;
+import javax.swing.filechooser.FileSystemView;
+import org.apache.commons.io.FileUtils;
+import org.openide.util.Exceptions;
+
+/**
+ * An S3 implementation of the FileSystemView interface to allow creation of
+ * JFileChoosers which view files from S3.
+ * 
+ * @author Will Winder
+ */
+public class S3FileSystemView extends FileSystemView {
+    private static final Logger logger = Logger.getLogger(S3FileSystemView.class.getName());
+    
+    // Must match "s3://<bucket>" or "s3://<bucket>/"
+    // Must match "s3:/<bucket>"  or "s3:/<bucket>/"
+    final static Pattern parseURI = Pattern.compile("^s3://?(?<bucket>[^/]+)/?(?<path>.*)$");
+    private static boolean hasBucket(String f) {
+        return parseURI(f).matches();
+    }
+    
+    private static Matcher parseURI(String f) {
+        return parseURI.matcher(f);
+    }
+    
+    final String id;
+    final String secret;
+    final MinioClient minioClient;
+    final List<Bucket> buckets;
+            
+    public S3FileSystemView(final String id, final String secret) throws Exception {
+        this.id = id;
+        this.secret = secret;
+        this.minioClient = new MinioClient("https://s3.amazonaws.com", id, secret);
+        
+        // Make sure it works with a simple call, cache buckets for fast init.
+        buckets = this.minioClient.listBuckets();
+    }
+
+    // Given an S3 URI, downloads the file to the target.
+    public void downloadFile(String uri, File target) {
+        Matcher m = parseURI(uri);
+        if (m.matches()) {
+            try {
+                InputStream s = minioClient.getObject(m.group("bucket"), m.group("path"));
```

Comment:
Is this stream getting closed? Add it to try block just in case:
`try(InputStream s = minioClient.getObject(m.group("bucket"), m.group("path"))) {`